src/CaptureClient/CaptureSaveToFileProcessor.cpp

```diff
@@ -0,0 +1,117 @@
+
+namespace {
+
+class CaptureSaveToFileProcessor : public CaptureEventProcessor {
+ public:
+  explicit CaptureSaveToFileProcessor(std::filesystem::path file_path,
+                                      absl::flat_hash_set<uint64_t> frame_track_function_ids,
+                                      std::function<void(const ErrorMessage&)> error_handler)
+      : frame_track_function_ids_(std::move(frame_track_function_ids)),
+        file_path_{std::move(file_path)},
+        error_handler_{std::move(error_handler)} {}
+  ~CaptureSaveToFileProcessor() override = default;
+
+  ErrorMessageOr<void> Initialize();
+  void ProcessEvent(const ClientCaptureEvent& event) override;
+
+ private:
+  ErrorMessageOr<void> WriteUserData();
+
+  absl::flat_hash_set<uint64_t> frame_track_function_ids_;
+  std::filesystem::path file_path_;
+  std::function<void(const ErrorMessage&)> error_handler_;
+  std::unique_ptr<CaptureFileOutputStream> output_stream_;
+};
+
+ErrorMessageOr<void> CaptureSaveToFileProcessor::Initialize() {
+  auto stream_or_error = CaptureFileOutputStream::Create(file_path_);
+  if (stream_or_error.has_error()) {
+    return ErrorMessage{absl::StrFormat("Failed to initialize CaptureSaveToFileProcessor: %s",
+                                        stream_or_error.error().message())};
+  }
+
+  output_stream_ = std::move(stream_or_error.value());
+
+  return outcome::success();
+}
+
+void CaptureSaveToFileProcessor::ProcessEvent(const ClientCaptureEvent& event) {
+  CHECK(output_stream_ != nullptr);
+  if (!output_stream_->IsOpen()) return;
+
+  auto write_result = output_stream_->WriteCaptureEvent(event);
+  if (write_result.has_error()) {
+    error_handler_(write_result.error());
+    return;
+  }
+
+  if (event.event_case() == ClientCaptureEvent::kCaptureFinished) {
+    // we are done - close the stream
+    auto close_result = output_stream_->Close();
+    if (close_result.has_error()) {
+      error_handler_(close_result.error());
+      return;
+    }
+
+    auto write_user_data_result = WriteUserData();
+    if (write_user_data_result.has_error()) {
+      error_handler_(write_user_data_result.error());
+      return;
+    }
+  }
+}
+
+ErrorMessageOr<void> CaptureSaveToFileProcessor::WriteUserData() {
+  OUTCOME_TRY(capture_file, CaptureFile::OpenForReadWrite(file_path_));
+  UserDefinedCaptureInfo user_defined_capture_data;
+  for (uint64_t function_id : frame_track_function_ids_) {
+    user_defined_capture_data.mutable_frame_tracks_info()->add_frame_track_function_ids(
+        function_id);
+  }
+
+  size_t message_size = user_defined_capture_data.ByteSizeLong();
+  auto buf = make_unique_for_overwrite<uint8_t[]>(message_size);
+  // SerializeToArray returns false if the size of the buffer is too small or too big
+  // https://github.com/protocolbuffers/protobuf/blob/5e84a6169cf0f9716c9285c95c860bcb355dbdc1/src/google/protobuf/message_lite.cc#L484
+  CHECK(user_defined_capture_data.SerializeToArray(buf.get(), message_size));
```

Comment:
Are we fine with temporarily serializing the full message in memory (I guess it's very small, isn't it?) and then writing to file?

Reply:
Yes, they are very small - we shouldn't have anything big here.